src/shapes/group.class.js

```diff
@@ -15,560 +21,872 @@
+          relativeToGroup ? this._onRelativeObjectAdded : this._onObjectAdded);
+        this._onAfterObjectsChange('added', Array.isArray(objects) ? objects : [objects]);
+      },
+
+      /**
+       * Remove objects
+       * @param {...fabric.Object} objects
+       */
+      remove: function () {
+        fabric.Collection.remove.call(this, arguments, this._onObjectRemoved);
+        this._onAfterObjectsChange('removed', Array.from(arguments));
+      },
+
+      /**
+       * Remove all objects
+       * @returns {fabric.Object[]} removed objects
+       */
+      removeAll: function () {
+        this._activeObjects = [];
+        var remove = this._objects.slice();
+        this.remove.apply(this, remove);
+        return remove;
+      },
+
+      /**
+       * backward compatibility
+       * @deprecated
+       */
+      addWithUpdate: function () {
+        this.add.apply(this, arguments);
+      },
+
+      /**
+       * backward compatibility
+       * @deprecated
+       */
+      removeWithUpdate: function () {
+        this.remove.apply(this, arguments);
+      },
+
+      /**
+       * invalidates layout on object modified
+       * @private
+       */
+      __objectMonitor: function (opt) {
+        this._applyLayoutStrategy(extend(clone(opt), {
+          type: 'object_modified'
+        }));
+        this._set('dirty', true);
+      },
+
+      /**
+       * keeps track of the selected objects
+       * @private
+       */
+      __objectSelectionMonitor: function (selected, opt) {
+        var object = opt.target;
+        if (selected) {
+          this._activeObjects.push(object);
+          this._set('dirty', true);
+        }
+        else if (this._activeObjects.length > 0) {
+          var index = this._activeObjects.indexOf(object);
+          if (index > -1) {
+            this._activeObjects.splice(index, 1);
+            this._set('dirty', true);
+          }
+        }
+      },
+
+      /**
+       * @private
+       * @param {boolean} watch
+       * @param {fabric.Object} object
+       */
+      _watchObject: function (watch, object) {
+        var directive = watch ? 'on' : 'off';
+        //  make sure we listen only once
+        watch && this._watchObject(false, object);
+        object[directive]('changed', this.__objectMonitor);
+        object[directive]('modified', this.__objectMonitor);
+        object[directive]('selected', this.__objectSelectionTracker);
+        object[directive]('deselected', this.__objectSelectionDisposer);
+      },
+
+      /**
+       * @private
+       * @param {fabric.Object} object
+       * @param {boolean} [removeParentTransform] true if object is in canvas coordinate plane
+       */
+      enterGroup: function (object, removeParentTransform) {
+        if (object.group) {
+          if (object.group === this) {
+            throw new Error('fabric.Group: duplicate objects are not supported inside group');
+          }
+          object.group.remove(object);
```

Comment:
Here I think it makes sense to preserve the old group's transform if `removeParentTransform` is true